library/rsa.c

```diff
@@ -528,16 +528,16 @@ int mbedtls_rsa_gen_key( mbedtls_rsa_context *ctx,
         MBEDTLS_MPI_CHK( mbedtls_mpi_gen_prime( &ctx->Q, nbits >> 1, 0,
                                                 f_rng, p_rng ) );
 
-        if( mbedtls_mpi_cmp_mpi( &ctx->P, &ctx->Q ) == 0 )
+        /* make sure the difference between p and q is not too small */
+        MBEDTLS_MPI_CHK( mbedtls_mpi_sub_mpi( &H, &ctx->P, &ctx->Q ) );
+        if( ( nbits >= 200 ) && ( mbedtls_mpi_bitlen( &H ) < (nbits >> 1) - 100 ) )
```

Comment:
I think the check should be `mbedtls_mpi_bitlen( &H ) < ( nbits >> 1 ) - 100 + 1`, because `mbedtls_mpi_bitlen( X ) >= m` entails `X >= 2^(m - 1)`. This, however, still leaves the highly unlikely possibility that `H == 2^(nbits / 2 - 100)`, which would violate the requirement `H > 2^(nbits / 2 - 100)` from the standard. To guard against rejection because of that -- should there ever be a meticulous check against FIPS-186-4 -- we might check for `mbedtls_mpi_bitlen( &H ) < ( nbits >> 1 ) - 100 + 2` instead.

Reply:
I used `mbedtls_mpi_bitlen( &H ) <= ( nbits >> 1 ) - 100 + 1` instead.